### ISSUE TITLE
UI automation

### ISSUE THREAD

**Issue by agowa** [microsoft/Windows-Containers#47]
Support UI (and UI automation) the windows ecosystem sadly is often "UI first" when it comes to software design. And the powershell/command line interfaces are not available or not documented at all. It would be great for sysadmins to be able to encapsulate these applications in containers, so that they are no longer bound to a single host and can be put into a k8s cluster without requiring something developers and Ranorex and a lot of time.

**Comment by vrapolinario** [comment]
Thanks for opening this issue @agowa338.
Have you checked the Windows base container image? (Not Server Core or Nano Server)
This image does provide support for UI automation - the downside comes with the size of the image, but the scenario should work.

**Comment by agowa** [comment]
Ok, than it might be documentation. I was looking for this for some time, but only found stuff like this though:

* [Sorry no. Microsoft specifically isn't supporting RDP or GUI's in containers, as far as I know. (...) but if you're trying to develop a desktop app, then containers aren't ideal.](https://github.com/BretFisher/ama/issues/72)
* [Microsoft made a change sometime after microsoft/windowsservercore:1709_KB4074588, breaking RDS.](https://withinrafael.com/2018/03/09/using-remote-desktop-services-in-containers/)
* [I’m looking forward to Windows apps being available with GUI; I think it’ll be really useful in my job.](https://forums.docker.com/t/how-to-run-a-gui-application-on-docker/48437/27)
* [UI testing is explicitly mentioned in the post, however I (and according to the reactions on that post some other people as well) are unable to actually interact with the app through WinAppDriver or other UI automation tools. (...) and also that there is no mouse support](https://github.com/MicrosoftDocs/Virtualization-Documentation-Private/issues/1455)

And this could also be related:
* [If you are using Windows 10 or Server 2016 your keyboard and mouse will not work on Session 0. (...) Additionally, Microsoft has removed the Interactive Services Detection Service on Windows 10 Version 1803.](https://kb.firedaemon.com/support/solutions/articles/4000106823-manually-enabling-interactive-services-interactive-service-detection-and-session-0)

**Comment by slavanap** [comment]
Same. We need official way for UI automation.

**Comment by slavanap** [comment]
We are waiting for updates here from Microsoft reps

**Comment by NikolayMatokh** [comment]
Still waiting....

**Comment by slavanap** [comment]
Microsoft team, please provide an update.

**Comment by agowa** [comment]
> Same. We need official way for UI automation.

From the replies this got I think some clarification is needed. I also think I now understand what @vrapolinario meant with there already being an official way. There is: https://docs.microsoft.com/en-us/dotnet/framework/ui-automation
However the API is quite "limited" in some while being quite capable in other situations.

That API is not able to fake input to notepad without the window being onscreen and having focus.
This is because notepad requires "System.Windows.Automation.TextPattern" to be used instead of "System.Windows.Automation.ValuePattern" and that one does not offer a set text method (more about why that matters in the example at https://docs.microsoft.com/de-de/dotnet/framework/ui-automation/add-content-to-a-text-box-using-ui-automation).
Therefore the only way to write into notepad automatically is using "System.Windows.Forms.SendKeys.SendWait". But if the focus suddenly changes, e.g. a popup opens, an error message is thrown, ... the input is provided to the wrong application...
Because of that, this is not a stable way to automate gui applications.

But there are also advantages of the current API. For example compared to AutoIt it allows to access elements within an WPF element within an dotnet application. And it doesn't require an additional runtime, powershell/dotnet is enough

Tl;Dr: The current state is better than I initially was aware of, but it still needs some polishing.

Example code for automating notepad:
```powershell
[void][System.Reflection.Assembly]::LoadWithPartialName("UIAutomationClient")
[void][System.Reflection.Assembly]::LoadWithPartialName("UIAutomationTypes")
[void][System.Reflection.Assembly]::LoadWithPartialName("UIAutomationProvider")
[void][System.Reflection.Assembly]::LoadWithPartialName("UIAutomationClientsideProviders")
[void][System.Reflection.Assembly]::LoadWithPartialName("System.Windows.Forms")

# According to the documentation this should be also called, but within powershell 7.1 it fails
# and within Windows Powershell (5.1) it is unstable and always throws an exception the first time it is called and succeeds the 2nd time...
# Also it doesn't seam to be needed for "just" automating notepad.
#$client = [System.Reflection.Assembly]::LoadWithPartialName("UIAutomationClientsideProviders")
#[Windows.Automation.ClientSettings]::RegisterClientSideProviderAssembly($client.GetName())

$autProc =  Start-Process notepad -PassThru
$rootElement = [Windows.Automation.AutomationElement]::RootElement
$condAUTProc = [Windows.Automation.PropertyCondition]::new([Windows.Automation.AutomationElement]::ProcessIdProperty, $autProc.Id)
$autElement = $rootElement.FindFirst([Windows.Automation.TreeScope]::Children, $condAUTProc)

# LocalizedControlTypeProperty is derived using "Accessibility Insights for Windows"
$documentElement = [Windows.Automation.PropertyCondition]::new([Windows.Automation.AutomationElement]::LocalizedControlTypeProperty, "document")
$document = $autElement.FindFirst([Windows.Automation.TreeScope]::Descendants, $documentElement)
$document.SetFocus()

[System.Windows.Forms.SendKeys]::SendWait("FooBar")
```

**Comment by vrapolinario** [comment]
Thanks @agowa338 for the follow up. Can I recommend you provide the feedback directly to the documentation above. It might be the case that the team who owns it has not seen feedback on it for Windows Containers.

@slavanap apologies for the delay. Does this helps with your case as well?

**Comment by slavanap** [comment]
Does 
https://github.com/microsoft/WinAppDriver
work in containers? Because if not then the suggested approach requires manual walkthrough and tinkering of every step, and doesn't support steps recording.

**Comment by vrapolinario** [comment]
There's an issue on that repo: https://github.com/microsoft/WinAppDriver/issues/1046 covering exactly that. While that tool appears not to work on Windows Containers as it depends on an active desktop session, there are alternatives in there.

I went ahead and opened an issue on their repo to add support for Windows Containers. Please remember this is a different team and we don't own their prioritization. We'll definitely work together if needed, though.

**Comment by slavanap** [comment]
The general issue with UI automation is that
* app installation,
* app startup sequence,
* app shutdown sequence
do involve iteration with GUI, although these legacy apps run as a backend service, they require interactions with GUI to get launched.

Also, for such apps state monitoring like an ability to take a screenshot would be really useful.

**Comment by grochoge** [comment]
UI automation appears to not work at all when I try using windows:1909 and TestStack.White.

After opening notepad.exe and using the underlying System.Windows.Automation, I don't see any children under the root element (Desktop). IE, the following prints nothing:
```c#
using System.Windows.Automation;
...
var lWalker = new TreeWalker( Condition.TrueCondition );
var lElement = lWalker.GetFirstChild( AutomationElement.RootElement );
while ( lElement != null )
{
    Console.Out.WriteLine( $"{lElement.GetType().Name} {lElement.Current.Name} {lElement.Current.ProcessId} {lElement.Current.BoundingRectangle}" );
    lElement = lWalker.GetNextSibling( lElement );
}
```

**Comment by grochoge** [comment]
The `Get-Process Notepad | ? MainWindowTitle -ne "" | Select MainWindowTitle` Powershell command mentioned as working with windows:1809 in another thread no longer works with windows:1909. 

What we could really use is a way to run multiple copies of Windows Sandbox as part of our CI. Why can't we use that as a Docker base? That also doesn't require a 6GB download/17GB image.

**Comment by agowa** [comment]
> Thanks @agowa338 for the follow up. Can I recommend you provide the feedback directly to the documentation above. It might be the case that the team who owns it has not seen feedback on it for Windows Containers.

The team apparently doesn't accept feedback. There isn't "provide feedback" at the bottom of the page as other docs pages have...

**Comment by vrapolinario** [comment]
Than can I recommend you provide the feedback directly at their repo? Here's the URL: https://github.com/microsoft/dotnet.

**Comment by slavanap** [comment]
Is it right that UI automation of legacy apps for Windows Containers infrastructure is not a priority for dev team right now?

**Comment by vrapolinario** [comment]
We made some progress on this with the Windows image that has a larger set of APIs available. Other than that, it seems the request is for tools that perform automation of UI tests need to support its own deployment on Windows containers. This unfortunately is out of our scope on the platform.

**Comment by grochoge** [comment]
@vrapolinario There appears to be negative progress on the Windows image that had a larger set of APIs available. From reading here, things that used to work in 1809 containers (like getting the main window title) no longer work in 1909 containers.

What we could really use is containers based on Windows Sandbox. But Microsoft puts artificial restrictions on that (along with the artificial licensing restrictions on normal Windows sessions) that makes it very very difficult to perform Windows Desktop UI automated testing.

**Comment by vrapolinario** [comment]
Adding @weijuans-msft for visibility.
Can't speak for the Sandbox option. We do want to address this scenario and we're looking into it, but no timeline at this point.
We'll update this issue once we have more to share.

**Comment by slavanap** [comment]
@vrapolinario re large set of APIs, is it now possible to take a screenshot of a running app in a Windows Container? If not, when will it be available?

**Comment by vrapolinario** [comment]
I'm not aware of a PowerShell option (or cmd for that matter) to take a screenshot, so I'm assuming it's not possible.

**Comment by vrapolinario** [comment]
Folks, after some internal investigation I'm getting the conclusion that we don't have a better answer at this point on UI automation, other than the APIs we provide on the Windows image. Since this seems to be a hot topic, I wanted to comment here and let you know that this is something we're taking in consideration as we evaluate new capabilities for Windows containers. Thanks for your input so far.

We'll close this issue for now, but please rest assured the inputs from this thread will feed our investments on upcoming releases.

**Comment by nchandra-seg** [comment]
Thanks for this. Please let us know if there is any roadmap

**Comment by david-terk** [comment]
With the release of this new image, are there any changes with respect to this topic?

https://techcommunity.microsoft.com/t5/containers/announcing-a-new-windows-server-container-image-preview/ba-p/2304897

**Comment by ghost** [comment]
update?

**Comment by pradeipp** [comment]
Hi @vrapolinario, any update on this? it's been 6+ months since you were last active in this thread.

**Comment by cwilhit** [comment]
Hey everyone, the Windows containers product team is still monitoring this issue. I am reopening the issue to keep for tracking the feature request, but **there are no new updates** to provide--it is still sitting on our backlog. If/when we do have an update, we will post it here or you can monitor the [Windows container blog](https://techcommunity.microsoft.com/t5/containers/bg-p/Containers).

**Comment by slavanap** [comment]
Up

**Comment by weijuans-msft** [comment]
At this time, we don't have any plan to support. Close it for now. Can reopen if this is changed.

**Comment by slavanap** [comment]
@weijuans-msft, with all the respect it's not completed, it is failed.

**Comment by nick-wellinghoff** [comment]
Can you offer an explanation as to why this use case is not supported? Its literally one of the most useful features of containers.  UI testing is a giant chore using VMs. Clearly all the code to make it work already exists (e.g. windows sandbox) and this is a totally artificial software imposed limit.

**Comment by xlazom00** [comment]
I did investigate this issue for a while and I did found that problem is in UIAutomation.
so https://github.com/microsoft/winappdriver don't work.
And if your testing framework use UIAutomation it don't work.

But I did add something like winappdriver directly into my app.
Selenium webdriver server so my application has selenium driver api.
And all works as expect. 
So if your target is to test own application inside windows containers you can do that.

I can see whole app with tree of items,....
I can click on items, edit text,....
I can see app with screenshots(selenium has api) 
window handles also working so you can see dialogs.
It is also working with system dialogs (open, save files, font picker,...)
I did add something like inspector into server so I can control whole app from browser.
